### PR TITLE
Allow FLAML_sample_size in starting_points

### PR DESCRIPTION
**Pull request by qingyun-wu**
- Allow "FLAML_sample_size" in starting_points 
- When "sample" is set False, users should not include "FLAML_sample_size" in the starting_points

### CODE REVIEW

File: flaml/automl.py [hunk 1]
```diff
@@ -2431,6 +2432,13 @@ def custom_metric(
                 < self._state.data_size[0]
             )
         )
+        if not self._sample and isinstance(starting_points, dict):
+            sample_size_exists = any(
+                [s.get("FLAML_sample_size") for s in starting_points.values() if s]
```
Comment (sonichi) [inline flaml/automl.py:2437]:
```suggestion
                ["FLAML_sample_size" in s for s in starting_points.values() if s]
```

File: flaml/automl.py [hunk 2]
```diff
@@ -2431,6 +2432,13 @@ def custom_metric(
...
+            )
+            assert (
+                not sample_size_exists
+            ), "When subsampling is disabled, should not include FLAML_sample_size in the starting point"
```
Comment (sonichi) [inline flaml/automl.py:2441]:
```suggestion
            ), "When subsampling is disabled, do not include "FLAML_sample_size" in the starting point"
```

File: test/automl/test_warmstart.py [hunk 3]
```diff
@@ -123,6 +123,35 @@ def test_nobudget(self):
         automl.fit(X_train, y_train)
         print(automl.best_config_per_estimator)
 
+    def test_FLAML_sample_size_in_starting_points(self):
+        from flaml.data import load_openml_dataset
+
+        X_train, X_test, y_train, y_test = load_openml_dataset(
+            dataset_id=1169, data_dir="./"
+        )
+        from flaml import AutoML
```
Comment (sonichi) [inline test/automl/test_warmstart.py:132]: move to line 128

File: test/automl/test_warmstart.py [hunk 4]
```diff
@@ -123,6 +123,35 @@ def test_nobudget(self):
...
+        print(len(y_train))
+        automl1.fit(X_train, y_train, **automl_settings)
+        print("automl1.best_config_per_estimator", automl1.best_config_per_estimator)
+
+        automl_settings["starting_points"] = automl1.best_config_per_estimator
+        automl2 = AutoML()
+        automl2.fit(X_train, y_train, **automl_settings)
+        automl_settings["sample"] = False
+        automl3 = AutoML()
+        automl3.fit(
```
Comment (sonichi) [inline test/automl/test_warmstart.py:149]: do you expect assertion error here? If so, need to catch it.

**Review by sonichi — CHANGES_REQUESTED** [review]

File: flaml/automl.py [hunk 5]
```diff
@@ -108,6 +108,7 @@ def valid_starting_point(self, starting_point, search_space):
                     value, search_space[name].get("domain")
                 )
                 for name, value in starting_point.items()
+                if name != "FLAML_sample_size"
```
Comment (sonichi) [inline flaml/automl.py:111]: It's not sufficient to exclude "FLAML_sample_size" from the starting_point. The specified sample size needs to be used as the starting sample size.

File: flaml/automl.py [hunk 6]
```diff
@@ -2431,6 +2432,18 @@ def custom_metric(
...
+            for _point in starting_points.values():
+                if _point and isinstance(_point, dict):
+                    _has_sample_size.append("FLAML_sample_size" in _point)
+                elif _point and isinstance(_point, list):
+                    _has_sample_size.append(
+                        any(["FLAML_sample_size" in c for c in _point])
+                    )
+            assert not any(
+                _has_sample_size
+            ), "When subsampling is disabled, do not include FLAML_sample_size in the starting point."
```
Comment (sonichi) [inline flaml/automl.py:2446]: No need to use the list _has_sample_size. Use the assert in the loop.

**Review by sonichi — COMMENTED** [review]
Not all the newly added code is covered by test.

**Comment by qingyun-wu** [comment]
Problem to fix: 
There are problems when starting points include points that have different values on the const HP (according to the built-in search space).

For example: when the starting_points is {'catboost': {'early_stopping_rounds': 10, 'learning_rate': 0.09999999999999996, 'n_estimators': 1, 'FLAML_sample_size': 10000}, 'lgbm': {'n_estimators': 16, 'num_leaves': 103, 'min_child_samples': 7, 'learning_rate': 1.0, 'log_max_bin': 8, 'colsample_bytree': 0.927858403859004, 'reg_alpha': 0.0009765625, 'reg_lambda': 0.1027216273706907, 'FLAML_sample_size': 40000}, }, while the domain of 'catboost' is a constant 8192 
 according to https://github.com/microsoft/FLAML/blob/main/flaml/model.py#L1545,
we may not find the index for catboost here: https://github.com/microsoft/FLAML/runs/7154070115?check_suite_focus=true#step:9:202

FYI @sonichi

File: flaml/automl.py [hunk 7]
```diff
@@ -102,12 +102,13 @@ def valid_starting_point_one_dim(self, value_one_dim, domain_one_dim):
         return True
 
     def valid_starting_point(self, starting_point, search_space):
-        return any(
+        return all(
             [
```
Comment (sonichi) [inline flaml/automl.py:106]: use generator expression for efficiency. remove []

File: flaml/automl.py [hunk 8]
```diff
@@ -657,8 +658,15 @@ def custom_metric(
                 add it as an additional element in the metric_constraints. Essentially 'pred_time_limit'
                 specifies a constraint about the prediction latency constraint in seconds.
             custom_hp: dict, default=None | The custom search space specified by user
-                Each key is the estimator name, each value is a dict of the custom search space for that estimator. Notice the
-                domain of the custom search space can either be a value of a sample.Domain object.
+                It is nested dict with keys being the estimator names, and values being dicts
```
Comment (sonichi) [inline flaml/automl.py:661]: add a period in the end of line 660.
nested dict -> a nested dict

File: flaml/automl.py [hunk 9]
```diff
@@ -657,8 +658,15 @@ def custom_metric(
...
+                of per estimator search space. In the per estimator search space dict,
```
Comment (sonichi) [inline flaml/automl.py:662]:
```suggestion
                per estimator search space. In the per estimator search space dict,
```

File: flaml/automl.py [hunk 10]
```diff
@@ -657,8 +658,15 @@ def custom_metric(
...
+                the keys are the hyperparameter names, and values are dicts of info ("domain",
+                "init_value", and "low_cost_init_value") about the search space associated with
+                the hyperparameter (i.e., per hyperparameter search space dict). When custom_hp
+                is provided, the built-in search space which is also a nested dict of per estimator
+                search space dict, will be updated with custom_hp. The update follows the vanilla
```
Comment (sonichi) [inline flaml/automl.py:667]: not sure what "vanilla nested dict update rule" means.

File: flaml/automl.py [hunk 11]
```diff
@@ -657,8 +658,15 @@ def custom_metric(
...
+                nested dict update rule. Note that the value for "domain" can either be a constant
+                or a sample.Domain object
```
Comment (sonichi) [inline flaml/automl.py:669]:
```suggestion
                or a sample.Domain object.
```

File: flaml/automl.py [hunk 12]
```diff
@@ -2419,18 +2427,68 @@ def custom_metric(
...
+                    and _point_per_estimator.get("FLAML_sample_size")
+                )
+                if sample_size:
+                    _sample_size_from_starting_points[_estimator] = sample_size
+                elif _point_per_estimator and isinstance(_point_per_estimator, list):
+                    _sample_size_set = set(
+                        [
+                            c["FLAML_sample_size"]
+                            for c in _point_per_estimator
+                            if "FLAML_sample_size" in c
```
Comment (sonichi) [inline flaml/automl.py:2446]:
```suggestion
                            for config in _point_per_estimator
                            if "FLAML_sample_size" in config
```

File: test/automl/test_warmstart.py [hunk 13]
```diff
@@ -1,7 +1,7 @@
 import unittest
 import numpy as np
 from sklearn.datasets import load_iris
-from flaml import AutoML
+from flaml import AutoML, automl
```
Comment (sonichi) [inline test/automl/test_warmstart.py:4]: why?

File: test/nlp/test_default.py [hunk 14]
```diff
@@ -57,15 +57,19 @@ def test_starting_point_not_in_search_space():
...
     automl_settings["starting_points"] = "data:test/nlp/default/"
     del automl_settings["fit_kwargs_by_estimator"][this_estimator_name]["model_path"]
 
     automl.fit(X_train, y_train, **automl_settings)
     assert (
-        len(automl._search_states[this_estimator_name].init_config) == 0
-    )  # check that init config is not updated, but search space is updated
+        len(automl._search_states[this_estimator_name].init_config)
+        == len(automl._search_states[this_estimator_name]._search_space_domain) - 3
+    ), "check that init config is not updated, but search space is updated"
```
Comment (sonichi) [inline test/nlp/test_default.py:72]: change the message accordingly

File: flaml/automl.py (hunk 10 above)
Comment (qingyun-wu) [reply flaml/automl.py]: Revised.

File: test/nlp/test_default.py (hunk 14 above)
Comment (qingyun-wu) [reply test/nlp/test_default.py]: Changed.

**Review by sonichi — APPROVED** [review]

File: test/nlp/test_default.py [hunk 15]
```diff
@@ -57,15 +57,23 @@ def test_starting_point_not_in_search_space():
...
     }
     automl_settings["starting_points"] = "data:test/nlp/default/"
     del automl_settings["fit_kwargs_by_estimator"][this_estimator_name]["model_path"]
 
     automl.fit(X_train, y_train, **automl_settings)
     assert (
-        len(automl._search_states[this_estimator_name].init_config) == 0
-    )  # check that init config is not updated, but search space is updated
+        len(automl._search_states[this_estimator_name].init_config)
+        == len(automl._search_states[this_estimator_name]._search_space_domain) - 3
```
Comment (liususan091219) [inline test/nlp/test_default.py:71]:
```suggestion
        == len(automl._search_states[this_estimator_name]._search_space_domain) - len(automl_settings["custom_hp"])
```

File: test/nlp/test_default.py [hunk 16]
```diff
@@ -57,15 +57,23 @@ def test_starting_point_not_in_search_space():
...
+    ), (
+        "The search space is updated with the custom_hp on 3 hyperparameters of "
```
Comment (liususan091219) [inline test/nlp/test_default.py:73]:
```suggestion
        "The search space is updated with the custom_hp on {} hyperparameters of "
```

File: test/nlp/test_default.py [hunk 17]
```diff
@@ -57,15 +57,23 @@ def test_starting_point_not_in_search_space():
...
+        "the specified estimator without an initial value. Thus a valid init config "
+        "should only contain the cardinality of the search space minus 3."
```
Comment (liususan091219) [inline test/nlp/test_default.py:75]:
```suggestion
        "should only contain the cardinality of the search space minus {}".format(len(automl_settings["custom_hp"]),  len(automl_settings["custom_hp"]))
```

**Review by liususan091219 — CHANGES_REQUESTED** [review]

File: test/nlp/test_default.py (hunk 17 above)
Comment (qingyun-wu) [reply test/nlp/test_default.py:75]: `len(automl_settings["custom_hp"]` should be `len(automl_settings["custom_hp"][this_estimator_name]` ?

**Review by liususan091219 — APPROVED** [review]

**Pull request merged**